StorageSystemModule for HPE OneView

ricardogpsf opened the pull request:
##### SUMMARY
Added new oneview_storage_system module for managing [HPE OneView Storage Systems](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.1/cic-api/en/api-docs/current/index.html#rest/storage-systems) resource and unit tests.

Related issue for more information: [HPE OneView support](https://github.com/ansible/ansible/issues/28354)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`oneview_storage_system`

##### ANSIBLE VERSION
```
ansible --version
ansible 2.5.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible-2.5.0-py2.7.egg/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Oct 10 2017, 02:49:49) [GCC 4.9.2]
```

##### ADDITIONAL INFORMATION
Note: to run the example you need an Oneview Appliance.
Example of usage:
```yaml
- name: Add a Storage System with one managed pool (before API500)
  oneview_storage_system:
    hostname: 172.16.101.48
    username: administrator
    password: my_password
    api_version: 300
    state: present
    data:
        credentials:
            ip_hostname: 172.18.11.12
            username: username
            password: password
        managedDomain: TestDomain
        managedPools:
          - domain: TestDomain
            type: StoragePoolV2
            name: CPG_FC-AO
            deviceType: FC

  delegate_to: localhost

- name: Add a StoreServ Storage System with one managed pool (API500 onwards)
  oneview_storage_system:
    hostname: 172.16.101.48
    username: administrator
    password: my_password
    api_version: 500
    state: present
    data:
      credentials:
          username: username
          password: password
      hostname: 172.18.11.12
      family: StoreServ
      deviceSpecificAttributes:
          managedDomain: TestDomain
          managedPools:
              - domain: TestDomain
                name: CPG_FC-AO
                deviceType: FC

  delegate_to: localhost

- name: Remove the storage system by its IP (before API500)
  oneview_storage_system:
    hostname: 172.16.101.48
    username: administrator
    password: my_password
    api_version: 300
    state: absent
    data:
        credentials:
            ip_hostname: 172.18.11.12

  delegate_to: localhost

- name: Remove the storage system by its IP (API500 onwards)
  oneview_storage_system:
    hostname: 172.16.101.48
    username: administrator
    password: my_password
    api_version: 500
    state: absent
    data:
        credentials:
            hostname: 172.18.11.12

  delegate_to: localhost
```

ricardogpsf committed:
Adds StorageSystemModule for HPE OneView

- Allows create, update and delete Storage System resources of HPE OneView
- Adds unit tests for it

mattclay commented:
CI failure in unit tests on Python 2.6, 3.5 and 3.6:

https://app.shippable.com/github/ansible/ansible/runs/43910/2/tests
https://app.shippable.com/github/ansible/ansible/runs/43910/4/tests
https://app.shippable.com/github/ansible/ansible/runs/43910/5/tests

fgbulsoni commented on lib/ansible/modules/remote_management/oneview/oneview_storage_system.py:
```diff
@@ -0,0 +1,233 @@
+                - C(present) will ensure data properties are compliant with OneView.
+                - C(absent) will remove the resource from OneView, if it exists.
+        choices: ['present', 'absent']
+        required: true
```
Should `default` to `present` instead of being `required`

fgbulsoni commented on lib/ansible/modules/remote_management/oneview/oneview_storage_system.py:
```diff
@@ -0,0 +1,233 @@
+    def __init__(self):
+        argument_spec = dict(
+            state=dict(
+                required=True,
```
Should be `default='present'` and this whole `state` should fit into one line.
I also suggest reordering the argument spec as mentioned in [here](https://github.com/ansible/ansible/pull/32798#discussion_r150491001)

fgbulsoni commented on test/units/modules/remote_management/oneview/test_oneview_storage_system.py:
```diff
@@ -0,0 +1,292 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+import yaml
+from ansible.compat.tests import unittest, mock
```
Suggestion: add a blank line here to separate `import X` and `from X import` types of import

ricardogpsf committed:
Changes documentation, organizes imports and default values of the HPE StorageSystemModule

fgbulsoni commented on lib/ansible/modules/remote_management/oneview/oneview_storage_system.py:
```diff
@@ -139,10 +139,7 @@ class StorageSystemModule(OneViewModuleBase):
-                required=True,
-                choices=['present', 'absent']
-            ),
+            state=dict(default='present', choices=['present', 'absent']),
```
This is missing the mandatory `type`.
As I mentioned before, I'd also recommend reorganizing everything inside the `argument_spec` as mentioned [here](https://github.com/ansible/ansible/pull/32798#discussion_r150491001).
as in: `state=(type='str', default='present', choices=['absent', 'present'])` .

Since `type` is mandatory, make it come first, then `default` or `required` and after that the rest.

ricardogpsf replied:
ah, sorry. My bad. =)

ricardogpsf committed:
Reorganizes the init method arguments of the HPE Oneview StorageSystemModule

fgbulsoni commented on lib/ansible/modules/remote_management/oneview/oneview_storage_system.py:
```diff
@@ -137,12 +137,13 @@ class StorageSystemModule(OneViewModuleBase):
-        )
-        super(StorageSystemModule, self).__init__(additional_arg_spec=argument_spec,
+        super(StorageSystemModule, self).__init__(additional_arg_spec=self.argument_spec,
                                                   validate_etag_support=True)
```
put `validate_etag_support=True)` in line `145`, there's no need for this split

fgbulsoni commented on lib/ansible/modules/remote_management/oneview/oneview_storage_system.py:
```diff
@@ -137,12 +137,13 @@ class StorageSystemModule(OneViewModuleBase):
     MSG_CREDENTIALS_MANDATORY = "The attribute 'credentials' is mandatory for Storage System creation."
 
+    argument_spec = dict(
+        state=dict(type='str', default='present', choices=['present', 'absent']),
```
choices=['absent', 'present']

ricardogpsf committed:
Reorders arguments from argument_spec in Oneview StorageSystemModule

fgbulsoni commented:
LGTM, shipit

adriane-cardozo commented:
shipit

ricardogpsf committed:
Uses Pytest to Oneview StorageSystemModule tests

- Updates tests to use pytest style
- Removes no_log param from the examples and adds parameter no_log=True for the Oneview base class

fgbulsoni commented on test/units/modules/remote_management/oneview/test_oneview_storage_system.py:
```diff
@@ -0,0 +1,300 @@
+    mock_ansible_module.exit_json.assert_called_once_with(
+        changed=False,
+        msg=StorageSystemModule.MSG_ALREADY_ABSENT
+    )
```
Looks to be missing the `if __name__ == '__main__':` part to be able to run through `python <name_of_file>`

fgbulsoni commented on test/units/modules/remote_management/oneview/test_oneview_storage_system.py:
```diff
@@ -0,0 +1,300 @@
+import yaml
+
+from ansible.compat.tests import mock
+from hpe_test_utils import mock_ov_client, mock_ansible_module
```
It is considered a bad practice to import pytest fixtures, I'm also seeing several flake errors commenting that these are being rewritten probably because of this.

fgbulsoni commented on test/units/modules/remote_management/oneview/hpe_test_utils.py:
```diff
@@ -15,12 +15,31 @@
 from hpOneView.oneview_client import OneViewClient
 
 
+@pytest.fixture
```
These fixtures should likely be inside `conftest.py`

ricardogpsf replied:
Good!

ricardogpsf committed:
Moves fixtures to conftest.py file for unit test of the Oneview StorageSystemModule

ricardogpsf committed:
Updates the documentation of examples of oneview_storage_system.py

fgbulsoni commented:
LGTM

ricardogpsf committed:
Updates the unit tests of StorageSystemModule to run tests of base test class

ricardogpsf committed:
Gets dynamically the current Oneview module being tested bu unit tests

tmiotto commented:
LGTM

ricardogpsf committed:
Adds a fixture to work like a setUp method to Oneview StorageSystemModule unit tests

ricardogpsf committed:
Moves setUp method of test to base class for StorageSystemModule tests

mattclay commented:
CI failure in unit tests:

```
>       self.resource = getattr(mock_ov_client, "%s" % (marker.args))
E       TypeError: not all arguments converted during string formatting

test/units/modules/remote_management/oneview/hpe_test_utils.py:31: TypeError
```

ricardogpsf commented:
@Akasurde I'm not working on it anymore.
I'am marking @soodpr and @madhav-bharadwaj because they could say something about of this task. Thanks.

Akasurde commented:
Hi @ricardogpsf, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
